website/project/views/node.py

```diff
@@ -824,6 +835,101 @@ def _get_children(node, auth, indent=0):
     return children
 
 
+def n_unread_comments(node, user, page, root_id=None, check=False):
+    """Return the number of unread comments on a node for a user."""
+    if not node.is_contributor(user):
+        return 0
+    if root_id is None or root_id == 'None' or page == 'node':
+        return n_unread_total(node, user, page, check=check)
+    root_target = Guid.load(root_id)
+    if root_target:
+        root_target = root_target.referent
+    else:
+        root_target = node.get_wiki_page(root_id, 1)
+    if page == 'files' and check:
+        exists, _ = check_file_exists(node, root_id)
+        if not exists:
+            return 0
+    default_timestamp = datetime(1970, 1, 1, 12, 0, 0)
+    view_timestamp = user.comments_viewed_timestamp.get(node._id, default_timestamp)
+    if isinstance(view_timestamp, dict):
+        view_timestamp = view_timestamp.get(page, default_timestamp)
+    if not page == 'node' and isinstance(view_timestamp, dict):
+        view_timestamp = view_timestamp.get(root_id, default_timestamp)
+    return Comment.find(Q('node', 'eq', node) &
+                        Q('user', 'ne', user) &
+                        Q('date_created', 'gt', view_timestamp) &
+                        Q('date_modified', 'gt', view_timestamp) &
+                        Q('is_deleted', 'eq', False) &
+                        Q('is_hidden', 'eq', False) &
+                        Q('root_target', 'eq', root_target)).count()
+
+
+def n_unread_total(node, user, page, check=False):
```

Comment:
There's a lot going on in this function. Are there opportunities for refactoring? Breaking up into smaller pieces?